File: scripts/extract_fields.py
#!/usr/bin/env python3
"""Extract schema-relevant fields from AdvisorHub article bodies.

Reads either wp-json post records (`research/wpjson/.../post_*.json`) or
raw HTML (`research/html/*.html`) and emits one JSONL row per article with
the fields surfaced by the regex/heuristic extractors below.

This is a starting point — production-grade extraction will need an LLM
pass for nuanced fields (named-entity disambiguation, allegation
categorization, etc.).
"""
import argparse
import json
import pathlib
import re
import sys
from bs4 import BeautifulSoup

MONEY_RE = re.compile(
    r"\$([\d,.]+)\s*(billion|bln|million|mln|m|b|k)?",
    re.I,
)
PCT_RE = re.compile(r"(\d{1,4}(?:\.\d+)?)\s*%")
YEAR_RE = re.compile(r"\b(19|20)\d{2}\b")
TENURE_RE = re.compile(r"(\d{1,2})-year (?:broker|veteran|advisor)", re.I)


def parse_money(s, unit):
    s = s.replace(",", "")
    if not s:
        return None
    v = float(s)
    if unit:
        u = unit.lower()
        if u.startswith("b"):
            v *= 1_000_000_000
        elif u.startswith("m"):
            v *= 1_000_000
        elif u == "k":
            v *= 1_000
    return v


def extract(text):
    out = {}
    moneys = [(parse_money(m.group(1), m.group(2)), m.group(0))
              for m in MONEY_RE.finditer(text)]
    out["money_mentions"] = [{"value": v, "phrase": p} for v, p in moneys if v]
    out["pct_mentions"] = [m.group(0) for m in PCT_RE.finditer(text)]
    out["years_mentioned"] = sorted(set(m.group(0) for m in YEAR_RE.finditer(text)))
    out["tenure_phrases"] = [m.group(0) for m in TENURE_RE.finditer(text)]

    # Crude advisor name capture: "First [M.] Last" near "advisor", "broker", "team"
    NAME_NEAR = re.compile(
        r"([A-Z][a-z]+(?:\s+[A-Z]\.)?\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)?)"
        r"(?=[^.]*?\b(?:advisor|broker|registered|managing director|vice president)\b)",
        re.I,
    )
    out["candidate_names"] = sorted(set(m.group(1) for m in NAME_NEAR.finditer(text)))

    # Firm mentions (small known set; expand iteratively)
    FIRMS = [
        "Morgan Stanley", "Merrill Lynch", "Merrill", "UBS", "Wells Fargo",
        "Rockefeller", "J.P. Morgan", "JPMorgan", "Goldman Sachs", "Stifel",
        "Raymond James", "RayJay", "LPL", "Ameriprise", "Edward Jones",
        "Cetera", "RBC", "First Republic", "Janney", "Hightower",
        "Beacon Pointe", "Focus Financial", "Steward Partners",
        "Wealthcare", "Hennion & Walsh", "Stanford Financial",
        "Chelsea Financial", "Smith Barney", "Lehman", "PaineWebber",
    ]
    out["firms_mentioned"] = sorted({f for f in FIRMS if f in text})

    # Sanctions
    fine_re = re.compile(r"fined\s+\$?([\d,]+)", re.I)
    susp_re = re.compile(r"suspend(?:ed)?\s+(?:for\s+)?(\w+|\d+)\s+(month|months|year|years)", re.I)
    bar_re = re.compile(r"\bbar(?:red)?\b.*?(?:from|by)?\s*([A-Z][a-zA-Z ]+)?", re.I)
    out["fines"] = [m.group(0) for m in fine_re.finditer(text)]
    out["suspensions"] = [m.group(0) for m in susp_re.finditer(text)]

    # Rule violations
    rule_re = re.compile(r"(?:FINRA |Finra )?Rule\s+(\d{3,5})", re.I)
    out["rule_violations"] = sorted(set(m.group(0) for m in rule_re.finditer(text)))

    # T-12 / recruiting deal %
    deal_re = re.compile(
        r"(\d{2,4})\s*%\s*(?:of\s+)?(?:trailing[- ]?(?:twelve|12)|T[- ]?12)",
        re.I,
    )
    out["recruiting_deal_pcts"] = [m.group(0) for m in deal_re.finditer(text)]

    return out


def from_wpjson(path):
    d = json.loads(pathlib.Path(path).read_text())
    html = d.get("content", {}).get("rendered", "")
    text = BeautifulSoup(html, "lxml").get_text("\n", strip=True)
    rec = {
        "source": "wpjson",
        "id": d.get("id"),
        "slug": d.get("slug"),
        "url": d.get("link"),
        "title": (d.get("title") or {}).get("rendered"),
        "date": d.get("date"),
        "modified": d.get("modified"),
        "categories": d.get("categories"),
        "tags": d.get("tags"),
        "coauthors": d.get("coauthors"),
    }
    rec["extracted"] = extract(text)
    rec["body_text"] = text
    return rec


def from_html(path):
    html = pathlib.Path(path).read_text()
    soup = BeautifulSoup(html, "lxml")
    h1 = soup.find("h1")
    title = h1.get_text(" ", strip=True) if h1 else ""
    pub = (soup.find("meta", {"property": "article:published_time"}) or {}).get("content", "") if soup.find("meta", {"property": "article:published_time"}) else ""
    paras = [p.get_text(" ", strip=True) for p in soup.find_all("p") if len(p.get_text(strip=True)) > 30]
    text = "\n".join(paras)
    rec = {
        "source": "html",
        "path": str(path),
        "title": title,
        "date": pub,
    }
    rec["extracted"] = extract(text)
    rec["body_text"] = text
    return rec


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--wpjson-dir", default="research/wpjson")
    ap.add_argument("--html-dir", default="research/html")
    ap.add_argument("--out", default="research/extracted.jsonl")
    args = ap.parse_args()
    out = pathlib.Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)
    n = 0
    with out.open("w", encoding="utf-8") as f:
        wp_dir = pathlib.Path(args.wpjson_dir)
        if wp_dir.exists():
            for p in list(wp_dir.rglob("post_*.json")) + list(wp_dir.rglob("*.wpjson.json")):
                rec = from_wpjson(p)
                f.write(json.dumps(rec, default=str) + "\n")
                n += 1
        html_dir = pathlib.Path(args.html_dir)
        if html_dir.exists():
            for p in html_dir.rglob("*.html"):
                rec = from_html(p)
                f.write(json.dumps(rec, default=str) + "\n")
                n += 1
    print(f"wrote {n} records to {out}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: src/scripts/extract_fields.ts
#!/usr/bin/env node
import { readdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";
import * as cheerio from "cheerio";

const FIRM_NAMES = [
  "Morgan Stanley",
  "Merrill Lynch",
  "UBS",
  "Wells Fargo",
  "Rockefeller",
  "J.P. Morgan",
  "Goldman Sachs",
  "Stifel",
  "Raymond James",
  "LPL",
  "Ameriprise",
  "Edward Jones",
  "Cetera",
  "RBC",
  "First Republic",
  "Janney",
  "Hightower",
  "Beacon Pointe",
  "Focus Financial",
  "Steward Partners",
  "Stanford Financial",
  "Chelsea Financial",
] as const;

const ROLE_PATTERN =
  /\b(?:advisor|broker|registered|managing director|vice president)\b/i;
const NAME_PATTERN =
  /\b[A-Z][a-z]+(?:\s+[A-Z]\.)?\s+[A-Z][a-z]+(?:\s+[A-Z][a-z]+)?\b/g;

/**
 * Returns strings in locale-aware alphabetical order.
 * @param values - Values to order for deterministic JSON output.
 * @returns A sorted copy of the supplied strings.
 */
function sortStrings(values: Iterable<string>): readonly string[] {
  return [...values].sort((a, b) => a.localeCompare(b));
}

/**
 * Parses money from source data.
 * @param raw - Raw source payload.
 * @param unit - Optional magnitude label such as million, billion, or k.
 * @returns The mention value in whole dollars, or null when the number is invalid.
 */
function parseMoney(raw: string, unit = ""): number | null {
  const value = Number(raw.replace(/,/g, ""));
  if (!Number.isFinite(value)) return null;
  if (/^b/i.test(unit)) return value * 1_000_000_000;
  if (/^m/i.test(unit)) return value * 1_000_000;
  if (/^k$/i.test(unit)) return value * 1_000;
  return value;
}

/**
 * Extracts lightweight entities and numeric facts from saved article text.
 * @param text - Plain text article body.
 * @returns Candidate facts used for manual data enrichment.
 */
export function extract(text: string): Record<string, unknown> {
  const moneyMentions = [
    ...text.matchAll(/\$([\d,.]+)\s*(billion|million|k)?/gi),
  ]
    .map(m => ({ value: parseMoney(m[1], m[2]), phrase: m[0] }))
    .filter(x => x.value);
  const sentencesWithRoles = text
    .split(/[.!?]\s+/)
    .filter(sentence => ROLE_PATTERN.test(sentence));
  const candidateNames = new Set(
    sentencesWithRoles.flatMap(sentence =>
      [...sentence.matchAll(NAME_PATTERN)].map(m => m[0])
    )
  );

  return {
    money_mentions: moneyMentions,
    pct_mentions: [...text.matchAll(/(\d{1,4}(?:\.\d+)?)\s*%/g)].map(m => m[0]),
    years_mentioned: sortStrings(
      new Set([...text.matchAll(/\b(?:19|20)\d{2}\b/g)].map(m => m[0]))
    ),
    tenure_phrases: [
      ...text.matchAll(/(\d{1,2})-year (?:broker|veteran|advisor)/gi),
    ].map(m => m[0]),
    candidate_names: sortStrings(candidateNames),
    firms_mentioned: sortStrings(FIRM_NAMES.filter(f => text.includes(f))),
    fines: [...text.matchAll(/fined\s+\$?([\d,]+)/gi)].map(m => m[0]),
    suspensions: [
      ...text.matchAll(
        /suspend(?:ed)?\s+(?:for\s+)?(\w+|\d+)\s+(month|months|year|years)/gi
      ),
    ].map(m => m[0]),
  };
}

/**
 * Extracts article facts from an HTML directory into a JSONL file.
 */
async function main(): Promise<void> {
  const out =
    process.argv[process.argv.indexOf("--out") + 1] ||
    "research/extracted.jsonl";
  const dir =
    process.argv[process.argv.indexOf("--html-dir") + 1] || "research/articles";
  const files = await readdir(dir);
  const lines = await Promise.all(
    files
      .filter(file => file.endsWith(".html"))
      .map(async file => {
        const html = await readFile(join(dir, file), "utf8");
        const $ = cheerio.load(html);
        const text = $.text().replace(/\s+/g, " ").trim();
        return JSON.stringify({ file, ...extract(text) });
      })
  );
  await writeFile(out, `${lines.join("\n")}${lines.length ? "\n" : ""}`);
}

if (import.meta.url === `file://${process.argv[1]}`) await main();
